src/relay/pass/fold_scale_axis.cc

```diff
@@ -0,0 +1,559 @@
+
+RELAY_REGISTER_OP("subtract")
+.set_attr<FForwardPrep>("FScaleAxisForwardPrep", AddSubForwardPrep);
+
+RELAY_REGISTER_OP("subtract")
+.set_attr<FForwardTransform>("FScaleAxisForwardTransform", AddSubForwardTransform);
+
+// Producer operators
+// Multiply produces the scale-axis pair.
+STuple MultiplyForwardTransform(const Call& ref_call,
+                                const AxesSet& expected_out_axes,
+                                const Array<STuple>& sargs) {
+  if (!expected_out_axes.defined()) return STuple();
+  // TODO(tvm-team) allow same axes accumulation
+  // not as important because it is less common in nn.
+  CHECK(!sargs[0]->axes.defined());
+  CHECK(!sargs[1]->axes.defined());
+  const auto* tlhs = ref_call->args[0]->checked_type().as<TensorTypeNode>();
+  const auto* trhs = ref_call->args[1]->checked_type().as<TensorTypeNode>();
+
+  Expr lhs = sargs[0]->value;
+  Expr rhs = sargs[1]->value;
+  auto rnode = make_node<STupleNode>();
+  if (MatchBroadcastToLeftAxes(tlhs, trhs, expected_out_axes, &rhs)) {
+    rnode->value = lhs;
+    rnode->scale = rhs;
+    rnode->axes = expected_out_axes;
+  } else if (MatchBroadcastToLeftAxes(trhs, tlhs, expected_out_axes, &lhs)) {
+    rnode->value = rhs;
+    rnode->scale = lhs;
+    rnode->axes = expected_out_axes;
+  }
+  return STuple(rnode);
+}
+
+RELAY_REGISTER_OP("multiply")
+.set_attr<FForwardTransform>("FScaleAxisForwardTransform", MultiplyForwardTransform);
+
+// Consumer operators
+// Conv2D send out requirement of axis folding.
+Array<AxesSet> Conv2DForwardPrep(const Call& call, AxesSet out) {
+  // TODO(tvm-team) support general data layout
+  // by transforming weight
+  const auto* param = call->attrs.as<Conv2DAttrs>();
+  CHECK(param != nullptr);
+  Layout data_layout(param->data_layout);
+  Layout weight_layout(param->weight_layout);
+  int c_big_axis = data_layout.indexof('C');
+  int c_small_axis = data_layout.indexof('c');
+  const auto* tdata = call->args[0]->checked_type().as<TensorTypeNode>();
+  CHECK(tdata) << "require checked type";
+
+  CHECK_GE(c_big_axis, 0);
+  AxesSet data_axes = NullValue<AxesSet>();
+  // For now, we only support simple pattern (no folded weight/data)
+  // More general layout can be supported under the current framework.
+  // By using a unified layout transformation.
+  // We only need to change the Prep and Mutate function.
+  //
+  // only handle depthwise or full conv2d.
+  // TODO(tvm-team) handle grouped conv by reshape + bcast
+  bool is_depthwise_conv2d =
+      is_const_int(tdata->shape[c_big_axis], param->groups);
+  if (weight_layout.indexof('i') < 0 &&
+      c_small_axis < 0 &&
+      (param->groups == 1 || is_depthwise_conv2d)) {
+    data_axes = {c_big_axis};
+  }
+  return {data_axes, NullValue<AxesSet>()};
+}
+
+// Conv2D consumes the scale axis during transformation.
+STuple Conv2DForwardTransform(const Call& ref_call,
+                                const AxesSet& expected_axes,
```

Comment:
`expected_axes` is unused. Should it be attached to `rnode`?

Reply:
also indent